lib/web_ui/lib/src/engine/layers.dart

```diff
@@ -924,23 +973,42 @@ class PlatformViewPathClip implements PlatformViewClip {
 }
 
 class LayerSliceBuilder {
-  factory LayerSliceBuilder() {
-    final (recorder, canvas) = debugRecorderFactory != null ? debugRecorderFactory!() : defaultRecorderFactory();
-    return LayerSliceBuilder._(recorder, canvas);
-  }
-  LayerSliceBuilder._(this.recorder, this.canvas);
-
   @visibleForTesting
-  static (ui.PictureRecorder, SceneCanvas) Function()? debugRecorderFactory;
+  static (ui.PictureRecorder, SceneCanvas) Function(ui.Rect)? debugRecorderFactory;
 
-  static (ui.PictureRecorder, SceneCanvas) defaultRecorderFactory() {
+  static (ui.PictureRecorder, SceneCanvas) defaultRecorderFactory(ui.Rect rect) {
     final ui.PictureRecorder recorder = ui.PictureRecorder();
-    final SceneCanvas canvas = ui.Canvas(recorder, ui.Rect.largest) as SceneCanvas;
+    final SceneCanvas canvas = ui.Canvas(recorder, rect) as SceneCanvas;
     return (recorder, canvas);
   }
 
-  final ui.PictureRecorder recorder;
-  final SceneCanvas canvas;
+  void addPicture(ui.Offset offset, ScenePicture picture) {
+    pictures.add((picture, offset));
+    final ui.Rect pictureRect = picture.cullRect.shift(offset);
+    cullRect = cullRect?.expandToInclude(pictureRect) ?? pictureRect;
+  }
+
+  LayerSlice buildWithOperation(LayerOperation operation) {
+    final ui.Rect recorderRect = operation.mapRect(cullRect ?? ui.Rect.zero);
+    final (recorder, canvas) = debugRecorderFactory != null ? debugRecorderFactory!(recorderRect) : defaultRecorderFactory(recorderRect);
```

Comment:
nit: maybe worth pulling the logic to create a recorder (using either the debug factory or the default factory) into its own method